proxy/src/proxy_metrics.rs

```diff
@@ -0,0 +1,188 @@
+//!
+//! Periodically collect proxy consumption metrics
+//! and push them to a HTTP endpoint.
+//!
+//!
+use std::{collections::HashMap, time::Duration};
+
+use tracing::{debug, error, log::info, trace};
+
+use consumption_metrics::{
+    ConsumptionMetricKind, ProxyConsumptionMetric, ProxyConsumptionMetricsKey, ProxyEventChunk,
+    CHUNK_SIZE,
+};
+
+use chrono::{DateTime, Utc};
+
+pub async fn collect_metrics(
+    metric_collection_endpoint: &reqwest::Url,
+    metric_collection_interval: Duration,
+    hostname: String,
+) -> anyhow::Result<()> {
+    scopeguard::defer! {
+        info!("collect_metrics has shut down");
+    }
+
+    let mut ticker = tokio::time::interval(metric_collection_interval);
+
+    info!(
+        "starting collect_metrics. metric_collection_endpoint: {}",
+        metric_collection_endpoint
+    );
+
+    // define client here to reuse it for all requests
+    let client = reqwest::Client::new();
+    let mut cached_metrics: HashMap<ProxyConsumptionMetricsKey, (u64, DateTime<Utc>)> =
+        HashMap::new();
+
+    loop {
+        tokio::select! {
```

Comment:
@funbringer 
could you please double check this code?

I'm not sure that I fully understand how `try_join_all` works.
Do I need to add some extra check for cancellation in this loop to ensure that it exits or it will happen automagically?

Reply:
Not sure about `try_join_all`, it's not present here. Could you elaborate?

Anyways, currently there's no special cancellation logic in the proxy, so that shouldn't be a problem.

Speaking of `select!`, you shouldn't need here because you're polling just one future. So it should be:

```rust
    loop {
        let _ = ticker.tick().await;

        let res = collect_metrics_task(
            &client,
            &mut cached_metrics,
            metric_collection_endpoint,
            hostname.clone(),
        )
        .await;

        match res {
            Err(e) => {
                error!("Failed to send consumption metrics: {} ", e);
            }
            Ok(_) => {
                trace!("collect_metrics_task completed successfully")
            }
        }
    }
```

However, let's take a closer look at the error handling part. Do you think we should terminate proxy if we are unable to collect or send metrics? I don't think that's a good idea, because clients are unlikely to be happy about that. We should consume the error and log it.

Reply:
sorry for the confusing question.
I was referring to [this code](https://github.com/neondatabase/neon/blob/9d6f88360eb4703ac42e293276307b083fd73e99/proxy/src/main.rs#L167) in in proxy main.rs
```
   // This will block until all tasks have completed.
    // Furthermore, the first one to fail will cancel the rest.
    let _: Vec<()> = futures::future::try_join_all(tasks).await?;
```

but `collect_metrics` future just loops forever and doesn't check for any shutdown signal.
Will it be somehow cancelled by tokio when other futures from the set fail, or I need to implement shutdown logic in `collect_metrics` so that main proxy thread would not hang forever?